### ISSUE TITLE
animal sniffer does not catch ByteBuffer's usages that break in Java 8 when compiled with Java 9+

### ISSUE THREAD

**Issue by voidzcy** [grpc/grpc-java#7360]
Originally reported in #7348. Existing usages are manually fixed by #7349. But we'd want to prevent this happens in the future.

> Java 9 introduces overridden methods with covariant return types for the following methods in java.nio.ByteBuffer:
> 
> - position​(int newPosition)
> - limit​(int newLimit)
> - flip​()
> - clear​()
> - mark​()
> - reset​()
> - rewind​()
> 
> In Java 9 they all now return ByteBuffer, whereas the methods they override return Buffer, resulting in exceptions like this when executing on Java 8 and lower:
> 
> java.lang.NoSuchMethodError: java.nio.ByteBuffer.limit(I)Ljava/nio/ByteBuffer
> 
> This is because the generated byte code includes the static return type of the method, which is not found on Java 8 and lower because the overloaded methods with covariant return types don't exist (the issue appears even with source and target 8 or lower in compilation parameters). The solution is to cast ByteBuffer instances to Buffer before calling the method.

We'd expect animal sniffer to catch usages that could potentially break Java 8 users.

**Comment by ejona86** [comment]
#7348 references https://github.com/protocolbuffers/protobuf/issues/7827#issuecomment-676563016 which references https://github.com/mojohaus/animal-sniffer/issues/77 . I was just going to wait for a fix, checking each release, or noticing an update notification.

**Comment by ogolberg** [comment]
note that my PR (https://github.com/mojohaus/animal-sniffer/pull/83) addressing this issue in animalsniffer has been open for a while - doesn't look like there's enough interest upstream to get this merged

at Toast, we're using a private fork of animalsniffer with the above fix applied

**Comment by cpovirk** [comment]
Yeah, based on your experience so far, I am likewise not too optimistic. I am hoping that maybe, if I post a beginning-to-end repro of the runtime failure, that will get some more traction on the PR. But we'll see.

**Comment by cpovirk** [comment]
(I should probably also mention why `--release`, their recommended alternative, isn't a great solution in our case.)

**Comment by ejona86** [comment]
@ogolberg, yeah, I had noticed it had been going on for a while. Although I had also seen the activity in July and figured in 1-2 more months I'd get more pessimistic, but I guess the useful discussion was with other users and not maintainers.

**Comment by cpovirk** [comment]
I gave it a shot in https://github.com/mojohaus/animal-sniffer/issues/77#issuecomment-681112257
